### PR TITLE
Enable displaying trace and span id on logs

### PR DESCRIPTION
**Pull request by BrynCooke**
To enable correlation between trace and logs trace_id and span_id should be included on the log messages.

 Json:
 ```
 {"timestamp":"2024-03-19T15:37:41.516453239Z","level":"INFO","trace_id":"54ac7e5f0e8ab90ae67b822e95ffcbb8","span_id":"9b3f88c602de0ceb","message":"Supergraph GraphQL response".....
 ```

 Text:
 ```
 2024-03-19T15:14:46.040435Z INFO trace_id: bbafc3f048b6137375dd78c10df18f50 span_id: 40ede28c5df1b5cc router{
 ```

 Configuration options have been added in logging for text and json formats defaulting to true for json and false for text.

Json:
```
   exporters:
     logging:
       stdout:
         format:
           json:
             display_span_id: true
             display_trace_id: true
```

Text:
```
  exporters:
    logging:
      stdout:
        format:
          text:
            display_span_id: true
            display_trace_id: true
```

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: docs/source/configuration/telemetry/exporters/logging/stdout.mdx [hunk 1]
```diff
@@ -140,29 +140,33 @@ telemetry:
...
+| `flavor`                    | `default`\|`compact`\|`pretty` | `default` | The overall output formatting.                                          |
+| `ansi_escape_codes`         | `true`\|`false`                | `false`   | Use ansi terminal escape codes.                                         |
+| `display_filename`          | `true`\|`false`                | `false`   | The filename where the log event was raised.                            |
+| `display_level`             | `true`\|`false`                | `true`    | The level of the log event, e.g. INFO, WARN, ERROR, TRACE.              |
+| `display_line_number`       | `true`\|`false`                | `false`   | The line number where the event was raised.                             |
+| `display_target`            | `true`\|`false`                | `false`   | The module name where the event was raised.                             |
+| `display_thread_id`         | `true`\|`false`                | `false`   | The id of the thread where the event was raised.                        |
```
Comment (shorgi) [inline docs/source/configuration/telemetry/exporters/logging/stdout.mdx:163]: Need global replace of `id` in body text to `ID`

```suggestion
| `display_thread_id`         | `true`\|`false`                | `false`   | The ID of the thread where the event was raised.                        |
```

File: docs/source/configuration/telemetry/exporters/logging/stdout.mdx [hunk 2]
```diff
@@ -140,29 +140,33 @@ telemetry:
...
+| `display_thread_name`       | `true`\|`false`                | `false`   | The name of the thread where the event was raised.                      |
+| `display_timestamp`         | `true`\|`false`                | `true`    | The timestamp of when the event was raised.                             |
+| `display_service_name`      | `true`\|`false`                | `false`   | The service name as configured in metrics common.                       |
+| `display_service_namespace` | `true`\|`false`                | `false`   | The service namespace as configured in metrics common.                  |
+| `display_trace_id`          | `true`\|`false`                | `false`   | The trace id of the span in which the event was raised.                 |
```
Comment (shorgi) [inline docs/source/configuration/telemetry/exporters/logging/stdout.mdx:168]:
```suggestion
| `display_trace_id`          | `true`\|`false`                | `false`   | The trace ID of the span in which the event was raised.                 |
```

File: docs/source/configuration/telemetry/exporters/logging/stdout.mdx [hunk 3]
```diff
@@ -140,29 +140,33 @@ telemetry:
...
+| `display_span_id`           | `true`\|`false`                | `false`   | The span id of the span in which the event was raised.                  |
```
Comment (shorgi) [inline docs/source/configuration/telemetry/exporters/logging/stdout.mdx:169]:
```suggestion
| `display_span_id`           | `true`\|`false`                | `false`   | The span ID of the span in which the event was raised.                  |
```

File: docs/source/configuration/telemetry/exporters/logging/stdout.mdx [hunk 4]
```diff
@@ -376,4 +386,7 @@ Example output with a list of spans:
 | `display_timestamp`   | `true`\|`false`   | `true`  | `timestamp`   | The timestamp of when the event was raised.                                            |
 | `display_span_list`   | `true`\|`false`   | `false` | `spans`       | A list of all spans to root in which the event was raised and all of their attributes. |
 | `display_resource`    | `true`\|`false`   | `false` | `resource`    | The resource as configured in tracing common.                                          |
+| `display_trace_id`    | `true`\|`false`   | `true`  | `trace_id`    | The trace id of the span in which the event was raised.                                |
```
Comment (shorgi) [inline docs/source/configuration/telemetry/exporters/logging/stdout.mdx:389]:
```suggestion
| `display_trace_id`    | `true`\|`false`   | `true`  | `trace_id`    | The trace ID of the span in which the event was raised.                                |
```

File: docs/source/configuration/telemetry/exporters/logging/stdout.mdx [hunk 5]
```diff
@@ -376,4 +386,7 @@ Example output with a list of spans:
...
+| `display_span_id`     | `true`\|`false`   | `true`  | `span_id`     | The span id of the span in which the event was raised.                                 |
```
Comment (shorgi) [inline docs/source/configuration/telemetry/exporters/logging/stdout.mdx:390]:
```suggestion
| `display_span_id`     | `true`\|`false`   | `true`  | `span_id`     | The span ID of the span in which the event was raised.                                 |
```

**Review by Geal — APPROVED** [review]
this looks fine but do we want to activate by default 2 new high cardinality fields in the logs? Wouldn't that surprise users?

**Review by bnjjj — APPROVED** [review]

**Comment by BrynCooke** [comment]
| this looks fine but do we want to activate by default 2 new high cardinality fields in the logs? Wouldn't that surprise users?

I see your point, but for instance New Relic and Datadog expect trace attributes to be on log messages, and the otel spec details that they should be there https://opentelemetry.io/docs/specs/otel/compatibility/logging_trace_context/#json-formats.

I think this'll give a better experience for the majority of users, and for logs carnality doesn't really matter.

**Comment by maschwenk** [comment]
@BrynCooke trying to understand how this PR interacts with this one:

- https://github.com/apollographql/router/pull/4764

**Comment by Samjin** [comment]
@BrynCooke How to work with existing `trace_id` with value that our service stacks already generated from upstream? Is there a way to reassign the existing value to otel generated trace_id so we can avoid multiple trace_ids, such as for all logging/trace/metrics? 

eg. I was trying to override `trace_id` in span with our own value, but it's expecting a boolean only. 
image

**Comment by Samjin** [comment]
> @Samjin I think you should use trace propagation for this use case https://www.apollographql.com/docs/router/configuration/telemetry/exporters/tracing/overview#propagation

Just tried it out locally with request header of trace_id, but it doesn't update the default OTEL `trace_id` value in span attribute and logs.

**Comment by Samjin** [comment]
> @Samjin I think you should use trace propagation for this use case https://www.apollographql.com/docs/router/configuration/telemetry/exporters/tracing/overview#propagation

Just tried it out with request header of `trace-id`, but it doesn't update the default OTEL `trace_id` value in span attribute and logs. Instead it did the opposite. It updated original request `trace-id` header and past to downstream.

**Comment by Samjin** [comment]
> @Samjin I think you should use trace propagation for this use case https://www.apollographql.com/docs/router/configuration/telemetry/exporters/tracing/overview#propagation

Create a ticket to override default trace_id. https://github.com/apollographql/router/issues/5056

**Pull request merged**